src/gui/qgsnewvectortabledialog.cpp

```diff
@@ -0,0 +1,753 @@
+{
+  mTableName->setText( name );
+}
+
+void QgsNewVectorTableDialog::setGeometryType( QgsWkbTypes::Type type )
+{
+  mGeomTypeCbo->setCurrentIndex( mGeomTypeCbo->findData( type ) );
+}
+
+void QgsNewVectorTableDialog::setCrs( const QgsCoordinateReferenceSystem &crs )
+{
+  mCrs->setCrs( crs );
+}
+
+QgsCoordinateReferenceSystem QgsNewVectorTableDialog::crs() const
+{
+  return mCrs->crs( );
+}
+
+QString QgsNewVectorTableDialog::tableName() const
+{
+  return mTableName->text();
+}
+
+QString QgsNewVectorTableDialog::schemaName() const
+{
+  return mSchemaCbo->currentText();
+}
+
+QString QgsNewVectorTableDialog::geometryColumnName() const
+{
+  return mGeomColumn->text();
+}
+
+QgsFields QgsNewVectorTableDialog::fields() const
+{
+  return mFieldModel ? mFieldModel->fields() : QgsFields();
+}
+
+QgsWkbTypes::Type QgsNewVectorTableDialog::geometryType() const
+{
+  QgsWkbTypes::Type type { static_cast<QgsWkbTypes::Type>( mGeomTypeCbo->currentData( ).toInt() ) };
+  if ( mHasMChk->isChecked() )
+  {
+    type = QgsWkbTypes::addM( type );
+  }
+  if ( mHasZChk->isChecked() )
+  {
+    type = QgsWkbTypes::addZ( type );
+  }
+  return type;
+}
+
+
+void QgsNewVectorTableDialog::setFields( const QgsFields &fields )
+{
+  if ( mFieldModel )
+  {
+    mFieldModel->setFields( fields );
+    validate();
+  }
+}
+
+QStringList QgsNewVectorTableDialog::validationErrors() const
+{
+  return mValidationErrors;
+}
+
+void QgsNewVectorTableDialog::updateButtons()
+{
+  mDeleteFieldBtn->setEnabled( mCurrentRow != -1 );
+  mFieldUpBtn->setEnabled( mCurrentRow != -1 && mCurrentRow != 0 );
+  mFieldDownBtn->setEnabled( mCurrentRow != -1 && mCurrentRow != fields().count() - 1 );
+}
+
+void QgsNewVectorTableDialog::selectRow( int row )
+{
+  QModelIndex index { mFieldsTableView->model()->index( row, 0 ) };
+  mFieldsTableView->setCurrentIndex( index );
+  QItemSelectionModel::SelectionFlags flags { QItemSelectionModel::ClearAndSelect | QItemSelectionModel::Current };
+  mFieldsTableView->selectionModel()->select( index, flags );
+  mFieldsTableView->scrollTo( index );
+}
+
+void QgsNewVectorTableDialog::validate()
+{
+  mValidationErrors.clear();
+
+  const bool isSpatial { mGeomTypeCbo->currentIndex() > 0 };
+  if ( mTableNames.contains( mTableName->text(), Qt::CaseSensitivity::CaseInsensitive ) )
+  {
+    mValidationErrors.push_back( tr( "Table '%1' already exists!" ).arg( mTableName->text() ) );
+  }
+  // Check for field names and geom col name
+  if ( isSpatial && fields().names().contains( mGeomColumn->text(), Qt::CaseSensitivity::CaseInsensitive ) )
+  {
+    mValidationErrors.push_back( tr( "Geometry column name cannot be equal to an existing field name!" ) );
+  }
+  // No geometry and no fields? No party!
+  if ( ! isSpatial && mFieldModel->fields().count() == 0 )
+  {
+    mValidationErrors.push_back( tr( "The table has no geometry column and no fields: cannot create and empty table!" ) );
```

Comment:
Spotted from your cast ;)
```suggestion
    mValidationErrors.push_back( tr( "The table has no geometry column and no fields: cannot create an empty table!" ) );
```